ImmutableComputableGraph code improvement + unit tests

mbabadi opened the pull request:
- reviewed and restricted the access modifiers of all ICG-related classes
- got rid of the functionality to hold on to old caches: whenever a cache goes out of date, the reference is immediately made null in the new ICG
- completely rewrote ComputableGraphStructure in a functional style
- got rid of unused and unnecessary methods
- Created an ImmutableComputableGraphUtils and factored out the builder and other common methods
- math equality asserts for NDArray
- unit tests for ComputableGraphStructure
- unit tests for ImmutableComputableGraph
- unit tests for ImmutableComputableGraphUtils

mbabadi commented:
@samuelklee can you suggest a reviewer? this is smaller than it seems :)

samuelklee commented:
@asmirnov239, perhaps you can take a look after you issue the PR for the germline WDL?  Otherwise I will take it when I get back.

asmirnov239 commented:
Yes sure! I can review it after I come back Wednesday next week

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
+     * Tests propagation of tags from descendents to parents
+     */
+    @Test(dataProvider = "allPossibleNodeFlags", invocationCount = NUM_TRIALS)
+    public void testTagPropagation(final boolean f_caching, final boolean f_external,
```
Shouldn't this belong to `ComputableGraphStructureUnitTest`?

mbabadi replied:
This test is actually redundant given the tests in `ComputableGraphStructureUnitTest` but I thought of keeping it because it is a _concrete_ test for the _concrete_ ICG worked out in this unit test. It may help some person in the future to understand what tag propagation means, etc.

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
+     * Asserts that the equality comparison of two {@link CacheNode}s is done just based on their key
+     */
+    @Test
+    public void testEquality() {
```
This test should go into the corresponding CacheNode unit test

asmirnov239 commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
-    @Test(enabled = false)
-    public void testPrimitiveUpdating() {
-        throw new AssertionFailedError("Test is not implemented yet");
+    @Test(expectedExceptions = UnsupportedOperationException.class)
```
Same here as above, and also for the tests below

droazen commented:
@mbabadi Migration instructions for this branch: https://github.com/broadinstitute/gatk/wiki/Migrating-branches-from-gatk-protected-to-gatk

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableCacheNode.java:
```diff
@@ -132,11 +119,11 @@ public ComputableCacheNode duplicate() {
      */
-    public ComputableCacheNode duplicateWithUpdatedValue(final Duplicable newValue) {
-        if (cacheEvals && newValue != null && !newValue.hasValue()) {
+    ComputableCacheNode duplicateWithUpdatedValue(final Duplicable newValue) {
```
good catch! done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
 /**
- * This class pre-computes a number of useful auxiliary properties for the DAG specified by
- * a set of {@link CacheNode}s. These include:
+ * This class performs consistency checks and computes several structural properties for the DAG specified by a
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
- * - topological order for evaluating all nodes associated to a tag (see {@link ImmutableComputableGraph}).
+ * - assertion for existence of no cycles
+ * - construction of maps of nodes to their descendants and ancestors
+ * - propagation of tags from descendants to ancestors
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+ * - topological order for evaluating a computable node
+ * - topological order for mutating a primitive/externally-computed node and updating the caches of all involved nodes
+ * - topological order for evaluating all nodes associated to a tag (see {@link ImmutableComputableGraph})
+ * - topological order for complete computation of the graph
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        nodeTagsSet = extractTags(nodeSet);
+        assertParentKeysExist(nodeSet, nodeKeysSet);
+
+        /* nodeKey -> set of parents' keys */
```
sounds good.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        /* nodeKey -> set of parents' keys */
+        parentsMap = getParentsMap(nodeSet);
+
+        /* nodeKey -> set of children's keys */
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        return nodeSet.stream().map(CacheNode::getKey).collect(Collectors.toSet());
+    }
+
+    private static Map<String, Integer> getTopologicalOrderMap(@Nonnull final Map<String,Set<String>> immediateParentsMap,
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        return topologicalOrderMap;
+    }
+
+    private static Map<Integer, Set<String>> getNodesByTopologicalOrderMap(Map<String, Integer> topologicalOrderMap,
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        IntStream.range(0, maxDepth + 1).forEach(depth ->
+                nodesByTopologicalOrderMap.put(depth,
+                        nodeKeysSet.stream().filter(node ->
+                                topologicalOrderMap.get(node) == depth).collect(Collectors.toSet())));
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
+        return allTagsMap;
+    }
+
+    private static Map<String, Set<String>> getNodesByInducedTagMap(@Nonnull final Map<String, Set<String>> allTagsMap,
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
 
-        /* topological order for evaluating all nodes associated to a tag */
-        topologicalOrderForTagEvaluation = new HashMap<>();
+    private static Map<String, List<String>> getTopologicalOrderForTagEvaluation(@Nonnull final Set<String> nodeTagsSet,
```
Yes, the same goes for node key. Perhaps I should do that...

mbabadi replied:
done. All Strings are now either `CacheNode.NodeKey` or `CacheNode.NodeTag`. Took a while!

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructure.java:
```diff
@@ -1,292 +1,379 @@
-                throw new CyclicGraphException("The graph has cycles");
-            }
-            depthsMap.replace(nodeKey, depth);
+     * */
```
fixed.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java:
```diff
@@ -59,50 +58,54 @@
- * (see below for updating caches selectively). This class throws an {@link IllegalStateException} if an old
- * cache is invoked in order to notify the user to update the cache manually.
+ * primitive or externally-computed node by invoking {@link ImmutableComputableGraphBuilder#withCacheAutoUpdate()}.
+ * This feature, however, is _not_ recommended as one may not need all the caches to be up-to-date at all times
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraph.java:
```diff
@@ -59,50 +58,54 @@
+ * The graph is queried either by calling {@link #fetchDirectly(String)} or
+ * {@link #fetchWithRequiredEvaluations(String)}. The former only fetches the values and throws an exception
+ * if the some of the required nodes are out of date, or a non-caching computable node is queried. The latter performs
+ * the required _necessary_ evaluations along the way.
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUtils.java:
```diff
@@ -0,0 +1,115 @@
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class ImmutableComputableGraphUtils {
```
That would restrict access to only classes in `org.broadinstitute.hellbender.tools.coveragemodel.cachemanager` and classes in `org.broadinstitute.hellbender.tools.coveragemodel` will not have access to builder methods, right?

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUtils.java:
```diff
@@ -0,0 +1,115 @@
+            cacheAutoUpdate = false;
+        }
+
+        public ImmutableComputableGraphBuilder primitiveNode(@Nonnull final String key,
```
done.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUtils.java:
```diff
@@ -0,0 +1,115 @@
+            return this;
+        }
+
+
```
fixed.

mbabadi commented on src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/CacheNode.java:
```diff
@@ -13,12 +13,20 @@
-
+abstract class CacheNode {
+    /**
+     * A string identifier for the cache node
```
you are welcome :D

mbabadi commented on src/main/java/org/broadinstitute/hellbender/utils/MathObjectAsserts.java:
```diff
@@ -16,6 +18,28 @@
 
     private MathObjectAsserts() {}
 
+    public static void assertNDArrayEquals(final INDArray actual, final INDArray expected,
```
because I'd like to have control over both relative and absolute errors. The INDArray implementation only deals with absolute errors.

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructureUnitTest.java:
```diff
@@ -1,19 +1,359 @@
+    }
+
+    @Test(invocationCount = NUM_TRIALS)
+    public void testNodeTagsAndKeys() {
```
done.

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructureUnitTest.java:
```diff
@@ -1,19 +1,359 @@
+    public void testAncestors() {
+        final RandomDAG dag = RandomDAG.getRandomDAG();
+        final ComputableGraphStructure cgs = new ComputableGraphStructure(dag.getEquivalentCacheNodeSet());
+        dag.nodeKeysSet.forEach(nodeKey -> Assert.assertTrue(dag.getAncestors(nodeKey).equals(cgs.getAncestors(nodeKey))));
```
I think smaller tests, if possible, are always preferred (for faster localization of failing tests).

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructureUnitTest.java:
```diff
@@ -1,19 +1,359 @@
+     * This test helper class creates a random DAG starting from a topological order. All helper methods are
+     * implemented in a brute-force manner.
+     */
+    private static final class RandomDAG {
```
Let me do this after the beta release -- I'm still hoping to get this PR to merged before beta!

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ComputableGraphStructureUnitTest.java:
```diff
@@ -1,19 +1,359 @@
+
+        private static Set<String> getRandomNodeKeys(final int maxNodes) {
+            return IntStream.range(0, rng.nextInt(maxNodes) + 1)
+                    .mapToObj(i -> "NODE_KEY_" + RandomStringUtils.randomAlphanumeric(NODE_KEY_LENGTH))
```
I put a condition to ensure that all keys are unique. Previously, the probability of collision was about `1e-49` (very passive aggressive :D).

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
+            "cos", UNARY_FUNCTION_COSINE,
+            "id", UNARY_FUNCTION_IDENTITY);
+
+    private static final List<String> EMPTY_STRING_LIST = new ArrayList<>();
```
done.

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
+    private static final Map<String, Duplicable> EMPTY_PARENTS = new HashMap<>();
+
+    /**
+     * Instructions for computing f(x, y) = F[2] ( F[0](x), F[1](x) )
```
yes, typo -- fixed.

mbabadi commented on src/test/java/org/broadinstitute/hellbender/tools/coveragemodel/cachemanager/ImmutableComputableGraphUnitTest.java:
```diff
@@ -1,91 +1,1138 @@
+        Assert.assertTrue(!icg_tmp.isValueDirectlyAvailable("h"));
+        assertIntactReferences(icg_0, icg_tmp, "y", "z", "g");
+
+        ImmutableComputableGraph icg_tmp_old = icg_tmp;
```
we discussed this

mbabadi commented:
@asmirnov239 OK, done with the review! :) I addressed all of your comments except for one (concrete example for graph structure). Will move to main repo and merge after the tests pass.
